src/coreclr/src/jit/morph.cpp

```diff
@@ -13707,15 +13712,31 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                         }
                         // If the type of the IND (typ) is a "small int", and the type of the local has the
                         // same width, then we can reduce to just the local variable -- it will be
-                        // correctly normalized, and signed/unsigned differences won't matter.
+                        // correctly normalized.
                         //
                         // The below transformation cannot be applied if the local var needs to be normalized on load.
                         else if (varTypeIsSmall(typ) && (genTypeSize(lvaTable[lclNum].lvType) == genTypeSize(typ)) &&
                                  !lvaTable[lclNum].lvNormalizeOnLoad())
                         {
-                            tree->gtType = typ = temp->TypeGet();
-                            foldAndReturnTemp  = true;
+                            // For any stores of small types, we will force loads to be normalized
+                            // this is necessary as we need to zero/sign extend any load
+                            // after this kind of store.
+                            //
+                            if (!isLoad)
+                            {
+                                varDsc->lvForceLoadNormalize = true;
+                            }
+                            // otherwise we have a load operation
+                            //
+                            // For loads signed/unsigned differences do matter.
+                            //
+                            if (varTypeIsUnsigned(lvaTable[lclNum].lvType) == varTypeIsUnsigned(typ))
```

Comment:
Should this be an `else if`? If not, I don't understand the "otherwise" comment above.

Reply:
Yes, I will change this to an else if,
I reordered these two during a refactoring

Reply:
A couple of additional methods had regressions: 

```
 21 (350.00% of base) : System.Private.CoreLib.dasm - System.Numerics.ConstantHelper:GetInt16WithAllBitsSet():short
 19 (316.67% of base) : System.Private.CoreLib.dasm - System.Numerics.ConstantHelper:GetSByteWithAllBitsSet():byte
```

Reply:
Additionally the BYTEMARK test IDEA also has a significant performace/codesize regression.

before: IDEA:  Iterations/sec: 5542.04724  Index: 84.76411
after:    IDEA:  Iterations/sec: 3428.97638  Index: 52.44527

Reply:
No other methods in the benchmarks had code diffs

Reply:
Did you have a look at the `IDEA` regression?

Reply:
Yes, it has many unsafe casts and is a awkward port of the C code:
The regression occurs in
  private static void cipher_idea(byte[] xin, byte[] xout, int offset, char[] Z)
Where we inline the MUL method:

```
                MUL(ref x1, Z[idx++]);
                MUL(ref x4, Z[idx++]);
                MUL(ref t2, Z[idx++]);
                MUL(ref t1, Z[idx++]);

    // These were macros in the original C code

    /* #define low16(x) ((x) & 0x0FFFF) */
    private static char low16(int x)
    {
        return (char)((x) & 0x0FFFF);
    }

    /* #define MUL(x,y) (x=mul(low16(x),y)) */
    private static void MUL(ref char x, char y)
    {
        x = mul(low16(x), y);
    }
```

Reply:
My other fix does better with a much smaller regression:

```
before: IDEA: Iterations/sec: 5542.04724 Index: 84.76411
after: IDEA:  Iterations/sec: 5244.63054  Index: 80.21521
```

Reply:
If we can change/fix the source code that another option:

Original:
```
    /* #define MUL(x,y) (x=mul(low16(x),y)) */
    private static void MUL(ref char x, char y)
    {
        x = mul(low16(x), y);
    }
```
Modification:
```
    /* #define MUL(x,y) (x=mul(low16(x),y)) */
    private static char MUL(char x, char y)
    {
        x = mul(low16(x), y);
    }
```

Reply:
I tried with the above code modification and with the change all version produce the same Assembly code
and the execution times are about the same as the original version.